pkg/localstore/sampler.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package localstore
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/bmtpool"
+	"github.com/ethersphere/bee/pkg/shed"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"go.uber.org/atomic"
+	"golang.org/x/sync/errgroup"
+)
+
+const sampleSize = 16
+
+var errDbClosed = errors.New("database closed")
+
+type sampleStat struct {
+	TotalIterated     atomic.Int64
+	NotFound          atomic.Int64
+	IterationDuration atomic.Int64
+	GetDuration       atomic.Int64
+	HmacrDuration     atomic.Int64
+}
+
+func (s *sampleStat) String() string {
+	return fmt.Sprintf(
+		"Total: %d NotFound: %d Iteration Durations: %d secs GetDuration: %d secs HmacrDuration: %d",
+		s.TotalIterated.Load(),
+		s.NotFound.Load(),
+		s.IterationDuration.Load()/1000000,
+		s.GetDuration.Load()/1000000,
+		s.HmacrDuration.Load()/1000000,
+	)
+}
+
+// ReserveSample generates the sample of reserve storage of a node required for the
+// storage incentives agent to participate in the lottery round. In order to generate
+// this sample we need to iterate through all the chunks in the node's reserve and
+// calculate the transformed hashes of all the chunks using the anchor as the salt.
+// In order to generate the transformed hashes, we will use the std hmac keyed-hash
+// implementation by using the anchor as the key. Nodes need to calculate the sample
+// in the most optimal way and there are time restrictions. The lottery round is a
+// time based round, so nodes participating in the round need to perform this
+// calculation within the round limits.
+// In order to optimize this we use a simple pipeline pattern:
+// Iterate chunk addresses -> Get the chunk data and calculate transformed hash -> Assemble the sample
+func (db *DB) ReserveSample(ctx context.Context, anchor []byte, storageDepth uint8) (storage.Sample, error) {
+
+	g, ctx := errgroup.WithContext(ctx)
+	addrChan := make(chan swarm.Address)
+	var stat sampleStat
+	logger := db.logger.WithName("sampler").V(1).Register()
+
+	// Phase 1: Iterate chunk addresses
+	g.Go(func() error {
+		defer close(addrChan)
+		iterationStart := time.Now()
+		for bin := storageDepth; bin <= swarm.MaxPO; bin++ {
+			err := db.pullIndex.Iterate(func(item shed.Item) (bool, error) {
```

Comment:
So I checked using the StartFrom option. The IterationTime is almost the same. The reason I kept it this way is we can further optimize this by having more workers here iterating on different bins.

Reply:
also the concept of swarm.MaxPO is an optimization in general, it is entirely possible that you might have chunks with PO's greater than the MaxPO (albeit very few). The correct thing to do is iterate on the whole thing.

Reply:
so if the performance is the same, then I'd go with iterating without bins

Reply:
And it may be approximately the same for small pullIndex populations, but it's a huge difference if you have lots of out-of-neighborhood chunks pinned in a node!   The fewer times we iterate the index, the better.